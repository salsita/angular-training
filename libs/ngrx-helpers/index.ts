export { NgrxHelpersModule } from './src/ngrx-helpers.module';
