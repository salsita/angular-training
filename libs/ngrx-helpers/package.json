{
  "$schema": "../../node_modules/ng-packagr/package.schema.json",
  "name": "@salsita/ngrx-helpers",
  "description": "ngrx helpers module for @salsita training",
  "version": "1.0.0",
  "peerDependencies": {
    "@angular/core": "5.2.6",
    "rxjs": "^5.5.6",
    "zone.js": "^0.8.19"
  },
  "ngPackage": {
    "lib": {
      "entryFile": "index.ts",
      "umdModuleIds": {}
    }
  }
}
