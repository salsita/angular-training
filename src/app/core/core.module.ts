import { HttpClientModule } from '@angular/common/http';
import { NgModule, Optional, SkipSelf } from '@angular/core';
import { ApiModule } from '@salsita/ng-api';
import { CrudModule } from '@salsita/ng-crud';
import { EntityRepositoryModule } from '@salsita/ng-entities';
import { RouterModule as RouterStoreModule } from '@salsita/ng-router';

import { environment } from '../../environments/environment';
import { SharedModule } from '../shared/shared.module';
import { ApiErrorToastComponent } from './api/api-error-toast/api-error-toast.component';
import { ApiLoaderComponent } from './api/api-loader/api-loader.component';

@NgModule({
  imports: [
    HttpClientModule,

    ApiModule.forRoot({ storeNamespace: 'api', baseUrl: environment.apiBaseUrl }),
    CrudModule.forRoot({ storeNamespace: 'crud' }),
    EntityRepositoryModule.forRoot({ storeNamespace: 'entityRepository' }),
    RouterStoreModule.forRoot({ storeNamespace: 'router' }),
    SharedModule
  ],
  declarations: [ApiErrorToastComponent, ApiLoaderComponent],
  exports: [ApiErrorToastComponent, ApiLoaderComponent, ApiModule, HttpClientModule]
})
export class CoreModule {
  constructor(
    @Optional()
    @SkipSelf()
    parentModule: CoreModule
  ) {
    if (parentModule) {
      throw new Error(
        `CoreModule has already been loaded. Import Core modules in the AppModule only.`
      );
    }
  }
}
