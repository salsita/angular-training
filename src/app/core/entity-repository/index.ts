export {
  EntityRepositoryModuleConfig,
  EntityRepositoryState,
  EntityRepositoryStateRoot,
  SingleEntityRepository
} from './entity-repository.interfaces';
export { EntityRepositorySelectors } from './entity-repository.selectors';
export { EntityRepositoryService } from './entity-repository.service';
export { EntityRepositoryModule } from './entity-repository.module';
