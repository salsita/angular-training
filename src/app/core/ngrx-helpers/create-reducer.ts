import { Action } from '@ngrx/store';

export const createReducer = <T, U extends Action>(handlers, initialState: T) => (
  state: T = initialState,
  action: U
): T => {
  const handler = handlers[action && action.type];
  return handler ? handler(state, action) : state;
};
