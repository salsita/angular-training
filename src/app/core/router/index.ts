export {
  FlatRouterStateSnapshot,
  RouterModuleConfig,
  RouterState,
  RouterStateRoot
} from './router.interfaces';
export { RouterSelectors } from './router.selectors';
export { CustomRouterStateSerializer } from './router.serializer';
export { RouterModule } from './router.module';
