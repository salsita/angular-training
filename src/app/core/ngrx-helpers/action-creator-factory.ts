import { Injectable } from '@angular/core';

/* Action Creator Factory
 * ======================
 * Creates typed action creators functions
 * const addTodo = ActionCreatorFactory.create<string>('ADD_TODO')
 *
 **/
export class ActionCreator<T> {
  constructor(public type: string = 'NOT_SET', public payload: T | null = null) {}
}
@Injectable()
export class ActionCreatorFactory {
  static create<T>(type: string, defaultPayloadValue: any = null) {
    return (payload: T | null = null): ActionCreator<any> => {
      const _payload = payload ? payload : defaultPayloadValue;
      return new ActionCreator<T>(type, _payload);
    };
  }

  create<T>(type: string, defaultPayloadValue: any = null) {
    return ActionCreatorFactory.create<T>(type, defaultPayloadValue);
  }
}
