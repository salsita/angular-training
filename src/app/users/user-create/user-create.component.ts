import { ChangeDetectionStrategy, Component } from '@angular/core';
import { Store } from '@ngrx/store';
import { Observable } from 'rxjs/Observable';
import { usersActionCreators } from '../+state/users.actions';
import { UsersSelectors } from '../+state/users.selectors';
import { Skill, User } from '../users.interfaces';

@Component({
  selector: 'app-user-create',
  templateUrl: './user-create.component.html',
  styleUrls: ['./user-create.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class UserCreateComponent {
  skills$: Observable<Skill[]>;

  constructor(private store: Store<any>, usersSelectors: UsersSelectors) {
    this.skills$ = usersSelectors.getSkills();
  }

  createUser(user: User) {
    this.store.dispatch(usersActionCreators.saveUser(user));
  }
}
