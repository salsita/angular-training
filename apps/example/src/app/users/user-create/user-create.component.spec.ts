// tslint:disable-next-line nx-enforce-module-boundaries
import { MockStore } from '@angular-training/ngrx-helpers/testing';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { ComponentFixture, TestBed, async } from '@angular/core/testing';
import { Store } from '@ngrx/store';

import { UsersSelectors } from '../+state/users.selectors';
import { MockUsersSelectors } from '../testing';
import { UserCreateComponent } from './user-create.component';

describe('UserCreateComponent', () => {
  let fixture: ComponentFixture<UserCreateComponent>;

  beforeEach(
    async(() => {
      TestBed.configureTestingModule({
        declarations: [UserCreateComponent],
        schemas: [NO_ERRORS_SCHEMA],
        providers: [
          { provide: UsersSelectors, useClass: MockUsersSelectors },
          { provide: Store, useValue: MockStore }
        ]
      }).compileComponents();
    })
  );

  beforeEach(() => {
    fixture = TestBed.createComponent(UserCreateComponent);
    fixture.detectChanges();
  });

  it('should render user form', () => {
    expect(fixture).toMatchSnapshot();
  });
});
