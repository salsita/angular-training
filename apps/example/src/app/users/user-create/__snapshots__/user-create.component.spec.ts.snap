// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserCreateComponent should render user form 1`] = `
<app-user-create
  skills$={[Function Observable]}
  store={[Function Function]}
>
  <app-user-form />
  
</app-user-create>
`;
